### PR TITLE
fix!: stop panics in parallel reactor processing

### PR DESCRIPTION
**Pull request by evan-forbes**
this stops some panics from occurring since the new queues were added on main.

I might have gone a bit overboard with the tests but given the importance here I think its worth it.

The only reason for the ! is that I cleaned up a function that returned an error when it didn't need to, and the only way to handle that error was to panic

### CODE REVIEW

File: p2p/base_reactor.go [hunk 1]
```diff
@@ -88,20 +87,34 @@ func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReac
...
 	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				fmt.Printf("processor panicked: %v\n", r)
+				// Try to stop the reactor gracefully
+				if err := base.Stop(); err != nil {
+					fmt.Printf("failed to stop reactor after panic: %v\n", err)
+				}
+			}
+		}()
```
Comment (evan-forbes) [inline p2p/base_reactor.go]: this catch is pretty redundant as the processor has the main catch

File: p2p/base_reactor.go [hunk 2]
```diff
@@ -86,24 +85,32 @@ func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReac
...
-			err = base.Stop()
-			if err != nil {
-				panic(err)
+		defer func() {
+			if r := recover(); r != nil {
+				fmt.Printf("processor panicked: %v\n", r)
+				// Try to stop the reactor gracefully
+				if err := base.Stop(); err != nil {
+					fmt.Printf("failed to stop reactor after panic: %v\n", err)
+				}
```
Comment (evan-forbes) [inline p2p/base_reactor.go]: this catch should never get hit but I figured why not

File: p2p/base_reactor.go [hunk 3]
```diff
@@ -160,64 +167,87 @@ func (br *BaseReactor) OnStop() {
...
-				}
-
-				msg := proto.Clone(mt)
+				// Process message with panic recovery for individual peer
+				process := func(ue UnprocessedEnvelope) error {
+					defer func() {
+						if r := recover(); r != nil {
+							disconnectPeer(baseReactor, ue.Src, fmt.Sprintf("panic in reactor: %v", r), impl.String())
+						}
+					}()
```
Comment (evan-forbes) [inline p2p/base_reactor.go]: the main catch is here, previously (in all existing networks!!) comet processed things serially and has a different catch in the connection since it doesn't protect against nil messages in other spots 😑 

the rest below is minor refactors

File: p2p/base_reactor.go [hunk 4]
```diff
@@ -86,24 +85,34 @@ func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReac
 		ctx:         ctx,
 		cancel:      cancel,
 		BaseService: *service.NewBaseService(nil, name, impl),
-		Switch:      nil,
+		Switch:      nil, // set by the switch later
 		incoming:    make(chan UnprocessedEnvelope, 100),
-		processor:   DefaultProcessor(impl),
+		processor:   nil, // Will be set after base is created
```
Comment (rach-id) [inline p2p/base_reactor.go:90]: why is this changed?

Comment (evan-forbes) [reply p2p/base_reactor.go:90]: good q

we need to pass the base reactor to the function, which is being created there.

before we were passing nil as that function didn't need to do anything with the reactor

File: consensus/propagation/have_wants.go [hunk 5]
```diff
@@ -112,6 +112,7 @@ func ReqLimit(partsCount int) int {
 }
 
 func (blockProp *Reactor) requestFromPeer(ps *PeerState) {
+	defer blockProp.ProtectPanic(ps.peer)
 	for {
```
Comment (evan-forbes) [inline consensus/propagation/have_wants.go:116]: added a method to the base reactor that we can should call in any goroutine that is handling input from a peer

this method is called in a goroutine elsewhere and while I don't think it could get hit, better to be safe

File: p2p/base_reactor.go [hunk 6]
```diff
@@ -160,64 +169,92 @@ func (br *BaseReactor) OnStop() {
 	br.cancel()
 }
 
-// DefaultProcessor unmarshalls the message and calls Receive on the reactor.
-// This preserves the sender's original order for all messages.
-func DefaultProcessor(impl Reactor) func(context.Context, <-chan UnprocessedEnvelope) error {
+// ProcessorWithReactor unmarshalls the message and calls Receive on the reactor.
+// This preserves the sender's original order for all messages and supports panic recovery with peer disconnection.
+func ProcessorWithReactor(impl Reactor, baseReactor *BaseReactor) func(context.Context, <-chan UnprocessedEnvelope) {
```
Comment (evan-forbes) [inline p2p/base_reactor.go:174]: this is not a required change, but we only have a single implementation therefore can only use a single function. tbh we could go further and make it a method

**Review by evan-forbes — COMMENTED** [review]
while the tests here do a good job of testing the new functionality, they don't have complete coverage for each reactor as there are still important footguns in using any goroutine.

therefore, we need to FLUP w/ a set of tests that submit literally every message as nil. It might even be better to write a fuzzer that will randomly submit nil messages to cover a larger range of possible states

File: p2p/base_reactor.go [hunk 7]
```diff
@@ -86,24 +85,34 @@ func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReac
...
+				fmt.Printf("processor panicked: %v\n", r)
+				// Try to stop the reactor gracefully only if it's running
+				if base.IsRunning() {
+					if err := base.Stop(); err != nil {
+						fmt.Printf("failed to stop reactor after panic: %v\n", err)
+					}
+				}
 			}
-		}
+		}()
```
Comment (cmwaters) [inline p2p/base_reactor.go:113]: Is this also for panics that may have been caused by messages for peers or something else?

Is this to make sure that the reactor stops which it wouldn't do during a normal panic?

File: p2p/base_reactor.go [hunk 8]
```diff
@@ -86,24 +85,34 @@ func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReac
...
 	go func() {
-		err := base.processor(ctx, base.incoming)
-		if err != nil {
-			fmt.Println("processor exited with error ", err)
-			err = base.Stop()
-			if err != nil {
-				panic(err)
+		defer func() {
+			if r := recover(); r != nil {
+				fmt.Printf("processor panicked: %v\n", r)
```
Comment (cmwaters) [inline p2p/base_reactor.go:105]: Would it be better to use the logger. Node operators might not be monitoring std out

**Review by cmwaters — APPROVED** [review]
I think this okay approach to take (as opposed to targeting specifically nil messages). My main concern would be that bugs in our code would just result in all a nodes peers being kicked which is a little more subtle than a panic

**Review by tzdybal — APPROVED** [review]
I think defensive approach here is good. Only one of those `recover` calls is caught by tests (if you comment out the `ProtectPanic` implementation, panic is still handled in base reactor).

File: p2p/base_reactor.go (hunk 8 above)
Comment (evan-forbes) [reply p2p/base_reactor.go:105]: agree makes sense I think we were just too lazy to pass the logger before, but will add to the flups

File: p2p/base_reactor.go (hunk 7 above)
Comment (evan-forbes) [reply p2p/base_reactor.go:113]: this is probably over the line for defensiveness as the processor needs to start protection asap, but this just makes sure that that occurs 

could definitely be removed, but I figured there's little harm in adding it
